crypto/serializer/serializer_lib.c

```diff
@@ -0,0 +1,154 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include <openssl/serializer.h>
+#include "internal/core.h"
+#include "internal/provider.h"
+#include "crypto/serializer.h"
+#include "crypto/evp.h"
+
+/*
+ * Support for OSSL_SERIALIZER_TYPE_to_bio:
+ * finding a suitable serializer
+ */
+
+struct selected_serializer_st {
+    OPENSSL_CTX *libctx;
+    const OSSL_PROVIDER *desired_provider;
+    const char *propquery;
+    /*
+     * Serializers offer two functions, one that handles object data in
+     * form of a OSSL_PARAM array, and one that directly handles a
+     * provider side object.  The latter requires that the serializer
+     * is offered by the same provider that holds that object, but is
+     * more desirable because it usually provides faster serialization.
+     *
+     * When looking up possible serializers, we save the first that can
+     * handle an OSSL_PARAM array in |first|, and the first that can
+     * handle a provider side object in |desired|.
+     */
+    OSSL_SERIALIZER *first;
+    OSSL_SERIALIZER *desired;
+};
+
+static void select_serializer(const char *name, void *data)
+{
+    struct selected_serializer_st *d = data;
+    OSSL_SERIALIZER *s = NULL;
+
+    /* No need to look further if we already have the more desirable option */
+    if (d->desired != NULL)
+        return;
+
+    if ((s = OSSL_SERIALIZER_fetch(d->libctx, name, d->propquery)) != NULL) {
+        if (d->first == NULL && s->serialize_data != NULL) {
+            d->first = s;
+        } else if (OSSL_SERIALIZER_provider(s) == d->desired_provider
+                   && s->serialize_object != NULL) {
+            OSSL_SERIALIZER_free(d->first);
+            d->first = NULL;
+            d->desired = s;
+        } else {
+            OSSL_SERIALIZER_free(s);
+        }
+    }
+}
+
+/*
+ * Support for OSSL_SERIALIZER_TYPE_to_bio:
+ * writing callback for the OSSL_PARAM (the implementation doesn't have
+ * intimate knowledge of the provider side object)
+ */
+
+struct serializer_write_data_st {
+    OSSL_SERIALIZER *serializer;
+    BIO *out;
+};
+
+static int serializer_write_cb(const OSSL_PARAM params[], void *arg)
+{
+    struct serializer_write_data_st *d = arg;
+    void *provctx =
+        ossl_provider_ctx(OSSL_SERIALIZER_provider(d->serializer));
+
+    return d->serializer->serialize_data(provctx, params, d->out);
+}
+
+/*
+ * OSSL_SERIALIZER_EVP_PKEY_to_bio() returns -2 if it can't find a suitable
+ * serializer.  This allows any caller to detect this and to use fallback
+ * methods, as opposed to when 0 is returned to signify that there was a
+ * fatal error.
+ */
+int OSSL_SERIALIZER_EVP_PKEY_to_bio(BIO *out, const EVP_PKEY *pkey,
+                                    const char *propquery, int want_domparams)
+{
+    EVP_KEYMGMT *keymgmt = pkey->pkeys[0].keymgmt;
+
+    if (keymgmt != NULL && pkey->pkeys[0].domainparams == want_domparams) {
```

Comment:
Hmm the want_domparams is an interesting check.. What if the key had domain params but we dont care that want_domparams = 0. Currently it returns -2?

Reply:
The way we (well, I) have structured it conceptually in providers make domain parameters a separate thing from keys.  Sure, our *implementation* uses the same structure for both, but still...

So yeah, if this particular `EVP_PKEY` holds domain parameters, and someone wants to get a key out of it, signalling that it's unsupported is the right thing to do...  I think.  Or should that be an error, i.e. return 0?

(the way the EVP API works, there actually is a separation between domain parameters and keys, even though subtle...  the separation visavi providers makes it more concrete)

Reply:
I've reworked this to look at the property query string and figure it out from there.

@paulidale, what I've done is kind of a hack, and I wonder if there's a better way to check if some specific property was queried using your property parser.

Reply:
I think something will need to be added to the property code to extract values.